runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -10283,45 +10269,111 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
          iCursor = generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, sizeReg,
                       generateS390MemoryReference(metaReg, offsetof(J9VMThread, heapAlloc), cg), iCursor);
       TR::LabelSymbol * fillerRemLabel = generateLabelSymbol(cg);
-      TR::LabelSymbol * doneLabel = generateLabelSymbol(cg);
 
       TR::LabelSymbol * fillerLoopLabel = generateLabelSymbol(cg);
 
-      // do this clear, if disableBatchClear is on
-      if (disableBatchClear && disableInitClear==NULL) //&& (node->getOpCodeValue() == TR::anewarray) && (node->getFirstChild()->getInt()>0) && (node->getFirstChild()->getInt()<6) )
+      if (inlineZeroInitialization)
          {
+         TR::Register * shiftReg = NULL;
+         TR::Register * addressReg = cg->allocateRegister();
+         if (conditions != NULL)
+            {
+            conditions->resetIsUsed();
```

Comment:
Why do we need to reset IsUsed again ?

Reply:
previous time it was on the condition of variable length only.

Reply:
Wouldn't this code be executed for both variable and constant array length ? 

This may be conservative way - i.e, those conditions are not attached to any instructions in between so may not have `isUsed` set to true that can cause the assert . Though, I think it would be OK if you reset the flag earlier in this function when you know that you will be adding additional registers to dependency conditions. 

For now I am ok with this as it is - but please when you are opening up PR with clean-up changes, address this one as well.

Reply:
I considered reseting it conditionally here (only if length is constant since it was done for variable) but it is faster to reset that flag even if it is already false than adding an if condition.
In the cleanup PR, I used scratch register manager that solves all the register and dependency issues.

Reply:
Keep in mind that scratch register manager will only manages the scratch registers - If a dependency has been attached before this part of code to instruction, and scratch register manager allocates new register (if it can not get one from the pool, it will allocate), than, you would need to reset. In the clean-up PR, please address how the passed dependency has been used to make sure.